### PR TITLE
Refactor challenge database with its tests

### PR DESCRIPTION
**Pull request by samymessouak**
## Summary
This Pull Request implements a new challenge feature in the Signify app — the *Chrono Challenge*. The changes include modifications to the database. The goal is to provide users with an engaging experience where they complete challenges against friends in the minimum time possible.

### 1. **Database Handling and Challenge Logic Modifications**
- Added new fields to the `Challenge` class:
  - Track player-specific round completion (`player1RoundCompleted`, `player2RoundCompleted`).
  - Track player times (`player1Times`, `player2Times`) for each round, recorded in milliseconds.
- Updated `ChallengeRepository` and `ChallengeRepositoryFirestore` to handle:
  - Creation, update, and retrieval of challenge states.
  - Saving player times and tracking each round's completion status.
  - Proper calculation and display of the player's total time after all rounds are completed.

## Checklist
- [x] Database modifications for tracking challenge states and player times.

Thank you for reviewing this pull request. Feel free to provide feedback or ask questions regarding any of the changes made.

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt [hunk 1]
```diff
@@ -14,34 +14,94 @@ class ChallengeRepositoryFireStore(private val db: FirebaseFirestore) : Challeng
       onSuccess: () -> Unit,
       onFailure: (Exception) -> Unit
   ) {
+    // Generate initial words for each round
+    val roundWords =
+        listOf("apple", "banana", "cherry") // Replace with dynamic word generation if needed
+
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt:20]: Consider dynamic word generation or sourcing from a database instead of hardcoding round words (apple, banana, cherry). This makes the solution more scalable and adaptable for future requirements.

File: app/src/test/java/com/github/se/signify/model/challenge/ChallengeTest.kt [hunk 2]
```diff
@@ -16,38 +16,45 @@ class ChallengeTest {
...
-    assertEquals("", challenge.currentGesture)
-    assertEquals(emptyMap<String, String>(), challenge.responses)
+    assertEquals(emptyList<Long>(), challenge.player1Times)
+    assertEquals(emptyList<Long>(), challenge.player2Times)
+    assertEquals(listOf(false, false, false), challenge.player1RoundCompleted)
+    assertEquals(listOf(false, false, false), challenge.player2RoundCompleted)
+    assertEquals("not_started", challenge.gameStatus)
   }
 
   @Test
```
Comment (nada-ezziti) [inline app/src/test/java/com/github/se/signify/model/challenge/ChallengeTest.kt:26]: There are some edge cases that are not taken into account in these tests : such as when both players complete a challenge simultaneously or a player leaves the challenge mid-way.

File: app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt [hunk 3]
```diff
@@ -14,34 +14,94 @@ class ChallengeRepositoryFireStore(private val db: FirebaseFirestore) : Challeng
...
+      playerId: String,
+      timeTaken: Long,
       onSuccess: () -> Unit,
       onFailure: (Exception) -> Unit
   ) {
-    db.collection(collectionPath)
+    db.collection("challenges")
         .document(challengeId)
-        .delete()
+        .update("playerTime", timeTaken)
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt:104]: Before updating Firestore, it's better to validate if timeTaken is a valid value (e.g., non-negative)

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
### Summary

This PR introduces the Chrono Challenge feature, which is a valuable addition to the Signify app, enhancing the user experience with an engaging and competitive gameplay mode. The implementation is well-thought-out and aligns with the app's goals, adding significant functionality to elevate user interaction.

While the feature is already strong, implementing a few minor refinements would enhance the overall quality and maintainability of the codebase. These small adjustments would ensure the feature is even more robust and scalable. Overall, this PR is an excellent contribution to the app and a step forward in creating a richer user experience.

File: app/src/test/java/com/github/se/signify/model/challenge/ChallengeTest.kt (hunk 2 above)
Comment (samymessouak) [reply app/src/test/java/com/github/se/signify/model/challenge/ChallengeTest.kt:26]: As I already have a 88% line coverage, I think i will focus on this for a next task.

File: app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt (hunk 3 above)
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt:104]: Nice catch

File: app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt (hunk 1 above)
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt:20]: As you can see with my comment, it was already taken into account and i was also going to do it for my next task, but nice catch it's necessary to do it!

**Review by SaFouNaldo — APPROVED** [review]
- The updates enhance functionality and add robustness to the challenge management system.
- Proper abstraction in the interface allows easy switching of backend implementations.
- Extensive test coverage provides confidence in the correctness of the implementation.
### Possible Improvements Needed:
- Avoid hardcoded default values in multiple places (e.g., roundWords).
-  Add comments explaining the new properties and methods, especially for recordPlayerTime and roundWords.
- Ensure that large datasets (e.g., thousands of challenges) won’t cause performance bottlenecks in Firestore queries or updates. A list that contains all the related player would have been useful here.

File: app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt (hunk 3 above)
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/challenge/ChallengeRepositoryFireStore.kt:105]: done

**Review by nada-ezziti — APPROVED** [review]
I will approve the PR for now, I hope the changes will be made in futur tasks like you said.
